### ISSUE TITLE
GitHub Actions for Lithops PRs and nightly builds - restore the missing functionality

### ISSUE THREAD

**Issue by gilv** [lithops-cloud/lithops#735]
Some of the recent commits had broken nightly builds support and enablement to use different buckets for CF and CE
Here are some comments that @Cohen-J-Omer provided:

- I’ve noticed my 2 config files were swapped with a single one and a few flags during runtime. The reason behind the separate config files for CE and CF was to assign a different bucket for each git-hub job, as I’ve noticed quite a bit of issues involving race conditions, due to simultaneous reads/writes on the same bucket.

- The nightly build workflow was completely removed. Indeed, currently the two workflows share a lot of code, but the nightly build was supposed to run more extensive checks in the future and send alerts to a slack channel. moreover, it was configured to run once every 24H, while the current one does not.

@Cohen-J-Omer is there anything else?
@erezh16 @JosepSampe

**Comment by JosepSampe** [comment]
>I’ve noticed my 2 config files were swapped with a single one and a few flags during runtime. The reason behind the separate config files for CE and CF was to assign a different bucket for each git-hub job, as I’ve noticed quite a bit of issues involving race conditions, due to simultaneous reads/writes on the same bucket.

Race conditions are produced if you run `lithops clean` command from multiple places while there is another job running in another places. Now `lithops clean` is only executed once at the end of the 2 tests, so there is no need to configure multiple buckets

>The nightly build workflow was completely removed. Indeed, currently the two workflows share a lot of code, but the nightly build was supposed to run more extensive checks in the future and send alerts to a slack channel. moreover, it was configured to run once every 24H, while the current one does not.

IMO there is no need to run the tests every 24 hours. We don't have anything critical in the project that needs to be tested every 24 hours. the code only changes on a PR, so run the test then is enough (I think). Moreover seems that this nightly build workflow is automatically activated on all forks, and me and others started to receive these emails:

![image](https://user-images.githubusercontent.com/10448186/127777276-f9807c65-6cc8-4437-876f-dc2ddc5dd3f8.png)

if in the future you plan to add more tests or critical tests, we can enable it, but for now, i think we don't need to activate this nightly build workflow (IMO).

**Comment by Cohen-J-Omer** [comment]
Hey @JosepSampe, good morning. 
> Race conditions are produced if you run lithops clean command from multiple places while there is another job running in another places. Now lithops clean is only executed once at the end of the 2 tests, so there is no need to configure multiple buckets

- My original idea was to make the git-action feature exclusive for the lithops source repository, 
  and the local test environment for the contributors.
  (which is why I left constants, that later would be swapped with the constants of the premium accounts). 

- As of right now, where users are given access to these features, I believe running `lithops clean` before starting to 
  run any other git-hub step is mandatory, since users might have previously stored runtimes on the bucket 
  (from previous jobs).   

- If you believe the only time a race condition on a bucket can occur when running `lithops clean` while another job is
  running, and you'd like to avoid creating a second bucket, I suggest reinstating the preliminary job you removed,
  where the bucket will be cleaned. Also, make all other jobs dependable upon said job (i.e., wait for its completion). 

> IMO there is no need to run the tests every 24 hours. We don't have anything critical in the project that needs to be tested every 24 hours. the code only changes on a PR, so run the test then is enough (I think). 

- You make a lot of sense, perhaps the nightly workflow could have been issued on a separate PR once a use for it was found,
  I agree. The thing is, the idea to start a test-run of the nightly build as soon as possible, was already decided in a 
  meeting, so we probably needed to forward your opinion before pulling the trigger on the change.

> Moreover, seems that this nightly build workflow is automatically activated on all forks, and me and others started to receive these emails:
-  I'm familiar with the matter. Once the git-hub workflow was to be operational (given the premium accounts) 
   I would add documentation, that among other things, would state that all forked repositories must either disable
   git-actions or allow only for select actions (through the repositories' settings). 

<ins>An issue I'd like to address</ins>:

I understand the faults in having a list of runnable jobs stored in a file (jobs_to_run.txt), that upon every change would
have to be re-committed, but I believe the current solution isn't future-proof to say the least:
* Currently, each time you'd like to omit a job from running you'd have to delete and re-create a rather 
long secret, which functions as a configuration file, that simply doesn't contain a specific segment (e.g. CE).
* In the future, you wouldn't be able to disable a configurations in that way. For example, lets say you'd like to disable CE-REDIS, but you'd like to run both CE-COS and CF-REDIS.
  You'd have to keep all segments of the configuration files intact, which will signal CE-REDIS to run.
      
I suggest reinstating the preliminary job in such a way that it will not only clean the bucket (as mentioned above), 
but will also read which jobs to run from a list of job names, perhaps residing in a new secret. 

As a side note, you missed an 'n' spelling 'code_engine' in main.yaml.

**Comment by JosepSampe** [comment]
>As of right now, where users are given access to these features, I believe running lithops clean before starting to
run any other git-hub step is mandatory, since users might have previously stored runtimes on the bucket
(from previous jobs).

There is no need to run `lithops clean` at the beginning of each job, if you run it at the end of all jobs in a workflow is enough, everything in the storage will be cleaned. Regarding the runtimes, I added 2 new steps that explicitly deletes the runtimes used in the execution of the workflow: [here](https://github.com/lithops-cloud/lithops/blob/master/.github/workflows/main.yml#L135) and [here](https://github.com/lithops-cloud/lithops/blob/master/.github/workflows/main.yml#L146). So in the next workflow execution everything is cleaned.

Basically I deleted the _nightly-build workflow_ because **for now** it had the exact same code of the _PR workflow_, and to avoid those errors emails to users. IMO receive those error messages every 24 hours is not good for the project and bad user experience. And if for now they don't provide any value, better deleted.

>Currently, each time you'd like to omit a job from running you'd have to delete and re-create a rather
long secret, which functions as a configuration file, that simply doesn't contain a specific segment (e.g. CE).
In the future, you wouldn't be able to disable a configurations in that way. For example, lets say you'd like to disable CE-REDIS, but you'd like to run both CE-COS and CF-REDIS.
You'd have to keep all segments of the configuration files intact, which will signal CE-REDIS to run

This is not like this. The main intention of allow avoid introducing some credentials in a long secret was not to enable/disable jobs trough the config, but allow the workflow run successfully if for some reason it is not possible to introduce some credentials of a specific service in the config. In my case, as a developer, it can be interesting having this workflow configured to run in my dev branch while I'm doing envelopment, so for example, if I don't want to run the code_engine job, I can now avoid introduce the credentials in my GITHUB secret without having to edit the source code of the workflow (yaml).

Regarding this _jobs_to_run.txt_, Why do you need it? I found it a bit messy. You created a script file located in another part of the project that reads this file located in another part of the project to decide which jobs to run. You can do exactly the same in the yaml itself without these 2 extra files. If you want to enable or disable jobs, just edit the workflow yaml file. This will be less messy, and easily understandable for others. In those yaml files you can do everything you want, there are thousands of options to use. For example, you can set env vars at the begging of the project to decide which jobs to run. [Take a look to what Ray guys do](https://github.com/ray-project/ray/blob/master/.github/workflows/main.yml), everything is put in the same yaml.
